torch/utils/data/dataloader.py

```diff
@@ -242,11 +286,25 @@ def _shutdown_workers(self):
         if not self.shutdown:
             self.shutdown = True
             self.done_event.set()
+            # if worker_manager_thread is waiting to put
+            if not self.data_queue.empty():
+                self.data_queue.get()
             for _ in self.workers:
                 self.index_queue.put(None)
+            # if all workers hang, no None is sent to worker_manager_thread, we
+            # put None to let worker_manager_thread exit
+            # empty check prevents put from hanging
+            if self.worker_result_queue.empty():
+                self.worker_result_queue.put(None)
+
+    def _remove_worker_pids_information(self):
+        if self.worker_pids_set:
+            _remove_worker_pids(id(self))
+            self.worker_pids_set = False
 
     def __del__(self):
         if self.num_workers > 0:
+            self._remove_worker_pids_information()
```

Comment:
Can you make it part of `_shutdown_workers()` the ordering is important and it's easy to forget to call this one every time. You can just inline it (the function has 3 lines)